Add Worldcat fetcher

GlennOlsson opened the pull request:
Add EntryBased fetcher for the Worldcat databse (#2581, #1065)

Problem right now is the API key. Tried requesting a sandbox key, but the request was denied. However I believe you being a trusted organisation might have better luck. You can request a key [at this link](https://www.oclc.org/developer/develop/authentication/how-to-request-a-wskey.en.html). The key can be added at [this line](https://github.com/JabRef/jabref/compare/master...augustjanse:worldcat-fetcher#diff-81925a19f0349951de113d912a0747a6R25)

Have tested the fetcher and importer with XML content from their [test api](https://platform.worldcat.org/api-explorer/apis/wcapi/Bib/OpenSearch), which has yield successful results

- [x] Change in CHANGELOG.md described (if applicable)a
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not: Issue created at <https://github.com/JabRef/user-documentation/issues>.

Siedlerchr commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ We refer to [GitHub issues](https://github.com/JabRef/jabref/issues) by using `#
 - Filenames of external files can no longer contain curly braces. [#5926](https://github.com/JabRef/jabref/pull/5926)
 - We made the filters more easily accessible in the integrity check dialog. [#5955](https://github.com/JabRef/jabref/pull/5955)
 - We reimplemented and improved the dialog "Customize entry types" [#4719](https://github.com/JabRef/jabref/issues/4719)
+- We added an Worldcat fetcher. [#1065](https://github.com/JabRef/jabref/issues/1065) [#2581](https://github.com/JabRef/jabref/issues/2581). [Worldcat](http://www.worldcat.org/webservices/catalog/)
```
```suggestion
- We added a WorldCcat fetcher. [#1065](https://github.com/JabRef/jabref/issues/1065) [#2581](https://github.com/JabRef/jabref/issues/2581). [WorldCat](http://www.worldcat.org/webservices/catalog/)
```

Siedlerchr commented on src/main/java/org/jabref/logic/importer/WebFetchers.java:
```diff
@@ -132,7 +133,8 @@ private WebFetchers() {
         set.add(new IsbnFetcher(importFormatPreferences));
         set.add(new MathSciNet(importFormatPreferences));
-        set.add(new CrossRef());
+		set.add(new CrossRef());
```
Indentation is off

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,87 @@
+	 * @return the earch query for the api
+	 */
+	private String getOpenSearchURL(String title){
+		String query = "&q=srw.ti+all+\"" + title.replaceAll(" ", "%20") + "\"";
```
Instead of manually replaying the illegal chars, I would create a URL (java.net.URL) and therefore let the java system handle the encoding of forbidden chars.

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,87 @@
+
+			return resp;
+		} catch (IOException e) {
+			e.printStackTrace();
```
Remove this printStackTrace please.
By wrapping the exception in a fetcher exception the exception will keep its stack trace and print it to the LOGGER

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,87 @@
+			return resp;
+		} catch (IOException e) {
+			e.printStackTrace();
+			throw new FetcherException("IO Exception", e);
```
Please a more meaningful exception title. e.g. Problem making a search request

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,87 @@
+	@Override
+	public List<BibEntry> performSearch(BibEntry entry) throws FetcherException {
+		Optional<String> entryTitle = entry.getLatexFreeField(StandardField.TITLE);
+		if(entryTitle.isPresent()){
```
Checkstyle violations. There must be spaces around the  braces. Please check the tests

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/WorldcatImporter.java:
```diff
@@ -0,0 +1,175 @@
+
+	public WorldcatImporter(){
+		//Used the same key as Worldcat fether
+		this.WORLDCAT_READ_URL = "http://www.worldcat.org/webservices/catalog/content/{OCLC-NUMBER}?recordSchema=info%3Asrw%2Fschema%2F1%2Fdc&wskey=" + WorldcatFetcher.API_KEY;
```
I saw it should be able to get JSON from the worldcat api:
Could you please test that?
That would improve parsing a lot (instead of using the old XML)
https://help.oclc.org/Discovery_and_Reference/WorldCat_Discovery/Troubleshooting/Using_the_WorldCat_Search_API_documentation_indicates_a_JSON_response_But_even_though_I_am_sending_an_Accept_header_with_the_value_application_JSON_I_keep_getting_the_XML_response

Siedlerchr commented on src/test/java/org/jabref/logic/importer/WorldcatImporterTest.java:
```diff
@@ -0,0 +1,56 @@
+import org.junit.jupiter.api.Test;
+
+public class WorldcatImporterTest {
+	private final String XML_WITH_RESULT = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><feed xmlns=\"http://www.w3.org/2005/Atom\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:diag=\"http://www.loc.gov/zing/srw/diagnostic/\" xmlns:oclcterms=\"http://purl.org/oclc/terms/\" xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:opensearch=\"http://a9.com/-/spec/opensearch/1.1/\"><title>OC ...
```
It's better to put such large files in a file and read it and compare it

tobiasdiez commented:
@koppor Could you please try to get an API key for us? That would be perfect. Thanks!

koppor commented:
@tobiasdiez I crafted a long answer yesterday, but it was lost. Short answer: Register --> login with university id --> I have to choose "Library of City Stuttgart" (or University of Stuttgart Library) (have to get an account for one of them, but that is doable) --> WorldCat then uses that Library --> I don't want all JabRef traffic going through the Stuttgart library --> Think, we need to add a key configuration and direct all users to https://www.oclc.org/developer/develop/authentication/how-to-request-a-wskey.en.html.

I would be OK for me to add a key for our JUnit tests.

Siedlerchr commented:
+1 for integration into the preferences

tobiasdiez commented:
@koppor Thanks for investigating this! This is a bit unfortunate as it decreases the user experience (and complexifies the code). Could you maybe contact them (I guess WCAPIRegister@oclc.org)? Maybe they provide access for us that is not bound to a specific university/library.

koppor commented:
Reply: JabRef can pay for a subscription. Alternatively, JabRef could ask a library to route traffic through them -- I would need to investigate on the costs for us. - In the current state, I would opt for a preference to enter the WorldCat key.

Siedlerchr commented:
@GlennOlsson  It would be nice if you could finish this PR.
Here are the things todo:
1. Add a field in the preferences to enter and store the API Key
2. Pass the preferences to the fetcher
3. Switch to the json API. I think it should be easier.

tobiasdiez commented:
Concerning the API key, I'm actually in contact with the TU Delft library. But that's a slow progress at the moment for the obvious reasons...

GlennOlsson commented:
Hi again and so sorry for the long delay! 

First off, JSON is not possible with this API. The link you provided @Siedlerchr ([this](https://help.oclc.org/Discovery_and_Reference/WorldCat_Discovery/Troubleshooting/Using_the_WorldCat_Search_API_documentation_indicates_a_JSON_response_But_even_though_I_am_sending_an_Accept_header_with_the_value_application_JSON_I_keep_getting_the_XML_response)) is regarding the ISBN search while I am implementing the Open Search API. You can see in their [API explorer](https://platform.worldcat.org/api-explorer/apis/wcapi/Bib/OpenSearch) that you can unfortunately not get a JSON response (test with `http://www.worldcat.org/webservices/catalog/search/worldcat/opensearch?q=APIs&wskey={built-in-api-key}&format=json` as the page you linked suggested). I agree that JSON is the easier method but it is sadly not possible.

I have now added a preference field which I had some trouble with, did not really get a grip on how the preferences were passed or where to put it, but decided to make an `APIKeyPreference` model if more API keys are being implemented this way. Added the preference field under the *Import* tab, also unsure where you would want it

Hopefully further improvements will go faster

Siedlerchr commented on src/main/java/org/jabref/preferences/JabRefPreferences.java:
```diff
@@ -775,8 +779,9 @@ public EntryEditorPreferences getEntryEditorPreferences() {
-                getBoolean(AVOID_OVERWRITING_KEY));
-    }
+				getBoolean(AVOID_OVERWRITING_KEY),
+				getApiKeyPreferences());
```
To make testing easier, it would be nice if you could directly pass the return value of `getWorldcatKey` (the string) to the fetcher.

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/WorldcatImporter.java:
```diff
@@ -0,0 +1,175 @@
+
+public class WorldcatImporter extends Importer {
+
+	private static final String DESCRIPTION = "Takes valid XML from Worldcat Open Search and parses them to BibEntry";
```
```suggestion
	private static final String DESCRIPTION = "Importer for Worldcat Open Search XML format";
```
Made it a bit consistent to the other importers.

GlennOlsson commented:
Sorry for the bad formatting. I am not using Intellij but was running the checkstyle gradle script which accepted my code, but it seem it is not checking tabs/spaces. Converted all files to spaces to comply with your standard

koppor commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,95 @@
+    private static final String WORLDCAT_OPEN_SEARCH_URL = "http://www.worldcat.org/webservices/catalog/search/opensearch?wskey=" + API_KEY;
+
+    public WorldcatFetcher (String worldcatKey) {
+        API_KEY = worldcatKey;
```
Codacy complains here: "Possible unsafe assignment to a non-final static field in a constructor.". I think, it's right and the static variable in line 29 should be replaced by an instance variable

Check https://app.codacy.com/gh/JabRef/jabref/pullRequest?prid=5083307#newIssuesView for details.

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -23,13 +23,14 @@
 public class WorldcatFetcher implements EntryBasedFetcher {
 
-    public static String API_KEY;
+    private String API_KEY;
```
```suggestion
    private final String apiKey;
```

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -23,13 +23,14 @@
     private static final String NAME = "Worldcat Fetcher";
 
-    private static final String WORLDCAT_OPEN_SEARCH_URL = "http://www.worldcat.org/webservices/catalog/search/opensearch?wskey=" + API_KEY;
+    private static String WORLDCAT_OPEN_SEARCH_URL = "http://www.worldcat.org/webservices/catalog/search/opensearch?wskey=";
```
```suggestion
    private String WORLDCAT_OPEN_SEARCH_URL = "http://www.worldcat.org/webservices/catalog/search/opensearch?wskey=";
```

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -23,13 +23,14 @@
 
-    public static String API_KEY;
+    private String API_KEY;
     private static final String NAME = "Worldcat Fetcher";
```
```suggestion

```

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,96 @@
+
+    @Override
+    public String getName () {
+        return NAME;
```
```suggestion
        return "Worldcat Fetcher";
```

GlennOlsson commented:
Alright, thanks!

tobiasdiez commented on CHANGELOG.md:
```diff
@@ -41,8 +41,11 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
+- We added a WorldCcat fetcher. [#1065](https://github.com/JabRef/jabref/issues/1065) [#2581](https://github.com/JabRef/jabref/issues/2581). [WorldCat](http://www.worldcat.org/webservices/catalog/)
 - We reimplemented and improved the dialog "Customize entry types". [#4719](https://github.com/JabRef/jabref/issues/4719)
 - We added an [American Physical Society](https://journals.aps.org/) fetcher. [#818](https://github.com/JabRef/jabref/issues/818)
+- We reimplemented and improved the dialog "Customize entry types" [#4719](https://github.com/JabRef/jabref/issues/4719)
```
This seems to be due to a merge conflict?

tobiasdiez commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,95 @@
+    private String WORLDCAT_OPEN_SEARCH_URL = "http://www.worldcat.org/webservices/catalog/search/opensearch?wskey=";
+
+    public WorldcatFetcher (String worldcatKey) {
+		this.apiKey = worldcatKey;
```
indent seems to be off

tobiasdiez commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,95 @@
+            String xmlResponse = makeOpenSearchRequest (entryTitle.get ());
+            WorldcatImporter importer = new WorldcatImporter (this.apiKey); 
+            ParserResult parserResult;
+            try { 
```
If you use the `EntryBasedParserFetcher` as the base class, then most of the error handling is done there (have a look for example at the `CrossRef` fetcher).

tobiasdiez commented on src/main/java/org/jabref/logic/importer/fileformat/WorldcatImporter.java:
```diff
@@ -0,0 +1,174 @@
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+public class WorldcatImporter extends Importer {
```
I think it suffices to implement the `Parser` interface. Importer are designed to import libraries from other bibliographic managers.

GlennOlsson commented:
Yes I understand your concern about the second wave of HTTP requests. These are made to gather further information about the entry like the date it was published. To do these we must first start to parse the XML do get the Worldcat id of each entry, and then do the request with those. So it was a design decision by me to do these requests in the importer which I am not sure about. Do you suggest that we parse the XML in both the fetcher and the importer? This seemed like extra work so I chose not to do that, but it you think that is better I can probably change it

I'll remove the duplicate in the CONTRIBUTORS file either way

tobiasdiez commented:
Yes, the idea would be to use the importer only to parse the final response. The whole progress of "user query > invoke api > parse xml > extract all ids > invoke api for each id" would be in the fetcher. 

For the importer you maybe can reuse https://github.com/JabRef/jabref/blob/19393bf1d00e4e0028436dc7b364fe3b7a7bfd81/src/main/java/org/jabref/logic/xmp/DublinCoreExtractor.java which apparently takes care of reading the dubline core format that is returned by WorldCat in the detailed query.

I leave it to you to decide if you want to improve the code in this regard or if you want to tackle the next issue.

GlennOlsson commented:
Alright so I've now changed so all HTTP requests are made in the Fetcher class. This will create a new XML document that is then sent to the importer for parsing

Hope this is what you meant

Siedlerchr commented on CHANGELOG.md:
```diff
@@ -50,8 +50,10 @@ Note that this project **does not** adhere to [Semantic Versioning](http://semve
+- We added a WorldCcat fetcher. [#1065](https://github.com/JabRef/jabref/issues/1065) [#2581](https://github.com/JabRef/jabref/issues/2581). [WorldCat](http://www.worldcat.org/webservices/catalog/)
 - We reimplemented and improved the dialog "Customize entry types". [#4719](https://github.com/JabRef/jabref/issues/4719)
 - We added an [American Physical Society](https://journals.aps.org/) fetcher. [#818](https://github.com/JabRef/jabref/issues/818)
+- We added an Worldcat fetcher. [#1065](https://github.com/JabRef/jabref/issues/1065) [#2581](https://github.com/JabRef/jabref/issues/2581). [Worldcat](http://www.worldcat.org/webservices/catalog/)
```
Seems like merging duplicated your description in the changelog

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,216 @@
+    private String WORLDCAT_OPEN_SEARCH_URL = "http://www.worldcat.org/webservices/catalog/search/opensearch?wskey=";
+    private String WORLDCAT_READ_URL = "http://www.worldcat.org/webservices/catalog/content/{OCLC-NUMBER}?recordSchema=info%3Asrw%2Fschema%2F1%2Fdc&wskey=";
+   
+
```
Checkstyle will fail: Two empty lines

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,216 @@
+     */
+    private Document parse (String s) {
+        try { 
+            BufferedReader r = new BufferedReader (new StringReader(s));
```
You never close the BufferedReader, this will create a ResourceLeak, wrap it in a try with resource statement
```
  try (BufferedReader br = new BufferedReader){
....
}
```

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,216 @@
+                //Transform XML to String
+                TransformerFactory tf = TransformerFactory.newInstance ();
+                Transformer t = tf.newTransformer ();
+                StringWriter sw = new StringWriter ();
```
Same as the other, you should wrap the writer in a try with resource block
```
try(StringWriter sw = new StringWriter()){
}
```

tobiasdiez commented on src/main/java/org/jabref/gui/preferences/ImportTabViewModel.java:
```diff
@@ -20,6 +20,8 @@
     private final StringProperty fileNamePatternProperty = new SimpleStringProperty();
     private final StringProperty fileDirPatternProperty = new SimpleStringProperty();
 
+    private final StringProperty apiKeyPatternProperty = new SimpleStringProperty();
```
There is no need for `pattern` here, right?

tobiasdiez commented on src/main/java/org/jabref/gui/preferences/ImportTab.fxml:
```diff
@@ -31,4 +31,20 @@
+            <RowConstraints vgrow="SOMETIMES"/>
+            <RowConstraints vgrow="SOMETIMES"/>
+        </rowConstraints>
+        <Label text="%Worldcat Fetcher Key" GridPane.rowIndex="1"/>
```
I guess `Fetcher Key` can be removed as this should be clear from the section header already.
It would be nice if could add a link to the site where the user can request the API key (or is this dependent on the library? in this case a "ask your library" comment is appropriate)

GlennOlsson replied:
Renamed it to Worldcat

I added a small description label with the link. However the labels are not selectable and I don't really know how to add a clickable hyperlink in this format either...

tobiasdiez replied:
You can use the `Hyperlink` controls for this (instead of a label). Have a look at the about dialog to do this.

GlennOlsson replied:
Coolio, fixed that!

tobiasdiez commented on src/main/java/org/jabref/logic/importer/fetcher/WorldcatFetcher.java:
```diff
@@ -0,0 +1,212 @@
+            String detailedXMLString;
+            try (StringWriter sw = new StringWriter ()){
+                //Transform XML to String
+                TransformerFactory tf = TransformerFactory.newInstance ();
```
We have the convention that there shouldn't be a space between the method name and the brackets, so `newInstance()`. Could you please reformat the code accordingly. Thanks

GlennOlsson replied:
Sorry, then I misunderstood the style schema! Fixed

tobiasdiez commented on src/test/java/org/jabref/logic/importer/fetcher/WorldcatFetcherTest.java:
```diff
@@ -0,0 +1,52 @@
+
+    @BeforeEach
+    public void setUp() {
+        fetcher = new WorldcatFetcher("API_KEY");
```
Can you please use the key `aMHOf2rfzUt3fuKkb7DXX8pkBv1AmcBWwwoSfwpt8CMhdUdxXscB4ESOmBPs4NlmYJmFtcSZ3Q5kMxzb`. This is a sandbox key that should work (for the build server and for local testing)

GlennOlsson replied:
Missed to resolve this, but this is the token being used

tobiasdiez reviewed (dismissed):
Thanks! The new tests are currently failing, is this expected and due to github? Please fix the localization and then this is ready to go from my side.

GlennOlsson commented:
Haha yeah, shit I missed the localization document, fixed now. I think the test might be failing because of the key, we get 407 which is [Proxy authentication required](https://httpstatuses.com/407), guessing there is some proxy used by Jabref for the tokens?

Siedlerchr commented:
I get the same error locally. There seems to be some problem with the key @tobiasdiez

GlennOlsson commented:
Yeah it's from [this reply](https://github.com/JabRef/jabref/issues/1065#issuecomment-246147573) back in the day. Tried with that during development but never got it to work then either. Maybe it has expired

tobiasdiez commented:
I contacted Worldcat. Let's see if this can be fixed quickly. If not, then we can disable the tests for now. Sorry @GlennOlsson for the lengthy process. Normally we try to get PRs merged more quickly.

GlennOlsson commented:
Don't worry about it! I am part of the problem with my code ;)

Siedlerchr commented:
@tobiasdiez  Any news from WorldCat regarding the key?

tobiasdiez commented:
Nope, sent them a follow-up...

koppor commented:
@GlennOlsson We would really like to see that this features comes into JabRef. May I ask whether you can find some time to finish this PR? It seems to be close to get it done. We have resources to provide timely feedback.

GlennOlsson commented:
I thought the only thing left was the Worldcat token? Didn’t realize there were things left to fix in the PR?

GlennOlsson commented:
The tests that fails are from the loss of token (from what I remember), as confirmed by @Siedlerchr

koppor commented:
Ah, sorry, was not quite reading the whole conversation.

I remember that probably every user will be asked to have a key (see https://github.com/JabRef/jabref/pull/6035#issuecomment-592395266). Their answer to me back then was similar to that (and that we just will have to pay if we want to have a "global" key). Maybe, I formulated our aims wrong and @tobiasdiez has more success.

Nevertheless, would it be possible to fix the merge conflicts? 😇

GlennOlsson commented:
Sorry, I have a hard time coming back to the project every few months to dive into the changes to fix the conflicts. I can probably fix them a final time when (if) we get a token but if I do it now I will probably have to dive into it again at a later point. I'm sorry, but due to the size of the project it's pretty tough to understand how everything has changed since the last time I saw it as I am not actively working on jabref

Sorry but I just don't have time to do this every now and then

koppor commented:
Side note: Working with fetchers is described at https://devdocs.jabref.org/advanced-reading/fetchers.

calixtus commented:
Hi @GlennOlsson, we discussed in our last dev call that we should somehow include this PR, so I took the liberty to fix the merge conflicts. I hope you don't mind.
However, I could not test this feature, since I do not have a Worldcat API key, but I made sure, that the given key is used (an error message can sometimes be a happy result too).

![grafik](https://user-images.githubusercontent.com/50491877/106438612-02746380-6477-11eb-9fb5-e67166e82a91.png) ![grafik](https://user-images.githubusercontent.com/50491877/106438692-18822400-6477-11eb-9541-da5c28f2d645.png)

![grafik](https://user-images.githubusercontent.com/50491877/106438787-3cde0080-6477-11eb-93c1-1adb2fae9bf1.png)

tobiasdiez commented:
I've contacted the TU Delft Library again and asked if we could use a key via them...let's see.

calixtus commented:
Remaining issue: The api seems to have changed. Requires a little bit more work.

koppor commented:
This PR currently implements WorldCat 1.0 API. This will be sunset at some time: https://www.oclc.org/developer/develop/web-services/worldcat-search-api.en.html. Currently, the version 2.0 is in test, and only "Libraries that maintain both WorldCat Discovery and OCLC Cataloging subscriptions." may use it.

koppor commented:
PDF on the WorldCat fetcher: https://matienzo.org/storage/2009/2009-Jun-WorldCatAPI.pdf

Example test:

Manual test using the WEb page: https://www.worldcat.org/search?q=ti%3AMarkdown+ti%3Aarchitectural+ti%3Adecision+ti%3Arecords%3A+ti%3AFormat+ti%3Aand+ti%3Atool+ti%3Asupport&qt=results_page

![grafik](https://user-images.githubusercontent.com/1366654/124836115-b88e2d00-df82-11eb-81a6-a3a0cdd347f3.png)

Similar call (only "Markdown Architectural") to the API:

https://www.worldcat.org/webservices/catalog/search/opensearch?wskey=...ourkey...&q=srw.ti%3DMarkdown%20and%20srw.ti%3DArchitectural

```xml
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<feed xmlns="http://www.w3.org/2005/Atom" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:diag="http://www.loc.gov/zing/srw/diagnostic/" xmlns:oclcterms="http://purl.org/oclc/terms/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:opensearch="http://a9.com/-/spec/opensearch/1.1/">
    <title>OCLC Worldcat Search: srw.ti=Markdown and srw.ti=Architectural</title>
    <id>http://worldcat.org/webservices/catalog/search/worldcat/opensearch?q=srw.ti%3DMarkdown+and+srw.ti%3DArchitectural&amp;start=1&amp;count=10&amp;format=atom&amp;wskey=...ourkey...</id>
    <updated>2021-07-07T18:17:09-04:00</updated>
    <subtitle>Search results for srw.ti=Markdown and srw.ti=Architectural at http://worldcat.org/webservices/catalog</subtitle>
    <opensearch:totalResults>1</opensearch:totalResults>
    <opensearch:startIndex>1</opensearch:startIndex>
    <opensearch:itemsPerPage>10</opensearch:itemsPerPage>
    <opensearch:Query role="request" searchTerms="srw.ti=Markdown and srw.ti=Architectural" startPage="1"/>
    <link rel="alternate" href="http://worldcat.org/webservices/catalog/search/worldcat/opensearch?q=srw.ti%3DMarkdown+and+srw.ti%3DArchitectural&amp;start=1&amp;count=10&amp;wskey=yuxeAzDLufk3pxfKFY2SJqUQg1yHGTwX7QkG2zpStsnTlSdhZISwQzdj3rQZ0Czl15y6ZuSMRRl4W2xV" type="text/html"/>
    <link rel="self" href="http://worldcat.org/webservices/catalog/search/worldcat/opensearch?q=srw.ti%3DMarkdown+and+srw.ti%3DArchitectural&amp;start=1&amp;count=10&amp;format=atom&amp;wskey=...ourkey..." type="application/atom+xml"/>
    <link rel="first" href="http://worldcat.org/webservices/catalog/search/worldcat/opensearch?q=srw.ti%3DMarkdown+and+srw.ti%3DArchitectural&amp;start=1&amp;count=10&amp;format=atom&amp;wskey=...ourkey..." type="application/atom+xml"/>
    <link rel="last" href="http://worldcat.org/webservices/catalog/search/worldcat/opensearch?q=srw.ti%3DMarkdown+and+srw.ti%3DArchitectural&amp;start=1&amp;count=10&amp;format=atom&amp;wskey=...ourkey..." type="application/atom+xml"/>
    <link rel="search" type="application/opensearchdescription+xml" href="http://worldcat.org/webservices/catalog/opensearch.description.xml"/>
    <entry>
        <author>
            <name>Abbott, Martin L.</name>
        </author>
        <title>The art of scalability : scalable web architecture, processes, and organizations for the modern enterprise</title>
        <link href="http://worldcat.org/oclc/904905227"/>
        <id>http://worldcat.org/oclc/904905227</id>
        <updated>2019-10-02T23:41:13Z</updated>
        <summary>In The Art of Scalability, Second Edition, leading scalability consultants Martin L. Abbott and Michael T. Fisher cover everything you need to know to smoothly scale products and services for any requirement. This extensively revised edition reflects new technologies, strategies, and lessons, as well as new case studies from the authors' pioneering consulting practice, AKF Partners. Writing for technical and nontechnical decision-makers, Abbott and Fisher cover everything that impacts scalability, including architecture, process, people, organization, and technology. Their insights and recommendations reflect more than thirty years of experience at companies ranging from eBay to Visa, and Salesforce.com to Apple. You'll find updated strategies for structuring organizations to maximize agility and scalability, as well as new insights into the cloud (IaaS/PaaS) transition, NoSQL, DevOps, business metrics, and more. Using this guide's tools and advice, you can systematically clear away obstacles to scalability-and achieve unprecedented IT and business performance.</summary>
        <dc:identifier>urn:ISBN:9780134032801</dc:identifier>
        <dc:identifier>urn:ISBN:0134032802</dc:identifier>
        <dc:identifier>urn:LCCN:2015009317</dc:identifier>
        <oclcterms:recordIdentifier>904905227</oclcterms:recordIdentifier>
    </entry>
</feed>
```

This is not at all a relation to the queried paper.

koppor commented:
The API key customization is done at https://github.com/JabRef/jabref/pull/7720.

koppor commented:
This PR is currently too much work to get it finished.

Main points:

1. Uses old WorldCat API 1.0. Not sure how future proof this is.
2. Querying using the 1.0 API lead to wrong results (see https://github.com/JabRef/jabref/pull/6035#issuecomment-875974870)
3. The PR misses the important feature to search for a query string (see https://github.com/JabRef/jabref/issues/1065#issuecomment-875947905). The current implemented functionality is to add BibTeX data to an existing entry.

Thus, I close this PR and hope that some other external contributor will take over.

Am sorry that I did not guide straight from the beginning through JabRef's different fetcher possibilities (the documentation at https://devdocs.jabref.org/advanced-reading/fetchers is far from being complete) and stressing that the most demanded feature is [that a user searches for new entries](https://docs.jabref.org/collect/import-using-online-bibliographic-database).
